File: 20230416_Sunday/20230416.cpp
// Leetcode: 1639. Number of Ways to Form a Target String Given a Dictionary.
// https://leetcode.com/problems/number-of-ways-to-form-a-target-string-given-a-dictionary/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 371 ms, faster than 38.80% of C++ online submissions for Number of
// Ways to Form a Target String Given a Dictionary.
// Memory Usage: 79 MB, less than 31.90% of C++ online submissions for Number of
// Ways to Form a Target String Given a Dictionary.
// 2023.04.16 Daily Challenge.
class Solution {
public:
    int numWays(vector<string>& words, string target) {
        // Copied from:
        // https://leetcode.com/problems/number-of-ways-to-form-a-target-string-given-a-dictionary/solution/
        const int alphabet = 26;
        const int mod = 1000000007;
        int m = target.size(), k = words[0].size();
        vector cnt(alphabet, vector<int>(k));
        for (int j = 0; j < k; j++) {
            for (const string& word : words) {
                cnt[word[j] - 'a'][j]++;
            }
        }
        vector dp(m + 1, vector<long long>(k + 1));
        dp[0][0] = 1;
        for (int i = 0; i <= m; i++) {
            for (int j = 0; j < k; j++) {
                if (i < m) {
                    (dp[i + 1][j + 1] += cnt[target[i] - 'a'][j] * dp[i][j]) %= mod;
                }
                (dp[i][j + 1] += dp[i][j]) %= mod;
            }
        }
        return dp[m][k];
    }
};

File: 20230416_Sunday/20230416.rb
# Leetcode: 1639. Number of Ways to Form a Target String Given a Dictionary.
# https://leetcode.com/problems/number-of-ways-to-form-a-target-string-given-a-dictionary/
# @param {String[]} words
# @param {String} target
# @return {Integer}
# TLE.
# 2023.04.16 Daily Challenge.
def num_ways(words, target)
  # Based on:
  # https://leetcode.com/problems/number-of-ways-to-form-a-target-string-given-a-dictionary/solution/
  alph,mod,m,k = 26, 1000000007, target.size, words[0].size
  cnt = Array.new(alph).map{ Array.new(k,0)}
  (0...k).each do |j|
    words.each do |word|
      cnt[word[j].ord - ?a.ord][j] += 1
    end
  end
  dp = Array.new(m+1).map{ Array.new(k+1,0)}
  dp[0][0] = 1
  (0..m).each do |i|
    (0...k).each do |j|
      if i < m
        a = cnt[target[i].ord - ?a.ord][j]
        dp[i+1][j+1] += a * dp[i][j] 
        dp[i+1][j+1] %= mod
      end
      dp[i][j+1] += dp[i][j]
      dp[i][j+1] %= mod
    end
  end
  return dp[m][k]
end
